### DISCUSSION TITLE
One UART shared by HSS and Linux

### DISCUSSION THREAD

**Discussion by rockoguido** [polarfire-soc/polarfire-soc-documentation#360] [HSS and U-Boot]
Hi.. 
working with a customer that has only one UART.  stage 0 bootloader using HSS and then that is to boot UBoot and Linux.
we have HSS surrendering the UART via UART_SURRENDER, the tricky part is getting UBoot and Linux shifted over to  UART0
I'm trying this on an Icicle Kit and just using UBoot as a test,  so trying to get UBoot to use Uart0
steps..

build using bitbake u-boot-mpfs
modified the device tree file (microchip-mpfs-icicle-kit.dts) to reflect the mapping of serial1 to uart0, by default serial 1 is mapped to uart1
bitbake u-boot-mpfs -c do_clean
bitbake u-boot-mpfs
deploy payload by sudo dd if=./payload.bin of=/dev/sda bs=1M
no effect, UBoot still using UART1

sidenote..
after I do a clean via bitbake the images directory under deploy is empty
After a build the images directory is repopulated BUT..  the timestamp on the images is not current, reflects the original build timestamp
??  do I need to clean more to truly reflect the changes made in the device tree file upon a rebuild??

Thx!
Dave

**Comment by eppidei** [comment]
Hi @rockoguido , I'm guessing here ,  probably the modified file is in the build directory and not "seen". What if you try to build with "-c compile -f" ?

**Comment by rockoguido** [comment]
[u-boot-mpfs.bb.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/12601971/u-boot-mpfs.bb.txt)

going to attach 4 files that resolved this issue.  2 recipe files and 2 corresponding patches, had to patch U-Boot and the Kernel to have both use Uart0.   1st 2 for UBoot, changed extension of files to allow to be inserted here

[0001-microchip-mpfs-icicle-kit-add-uart0-as-console patch.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/12601978/0001-microchip-mpfs-icicle-kit-add-uart0-as-console.patch.txt)

next 2 for Linux kernel



[mpfs-linux.bb.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/12601986/mpfs-linux.bb.txt)


[0001-dts-mpfs-icicle-kit-console-to-uart0.patch.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/12601990/0001-dts-mpfs-icicle-kit-console-to-uart0.patch.txt)

**Comment by rockoguido** [comment]
hi.  have a follow-on related question.    For me on the Icicle Kit the single UART scenario works fine.    The HSS boots UBoot and then surrenders the UART.   of course at this point the CLI in HSS is not available.   

For the customer they also require the single UART while booting HSS -> UBoot -> Linux.   We are hung up for them on getting Linux to boot, console just goes dark post UBoot message "Booting Linux"  

Are trying a few things related to memory management since customer has only 1Gig of DDR as opposed to the Icicle having 2 Gig.
Have reduced the DDR amount in HSS along with updates in the Kernel Device tree, which we have verified.    So all updates work on the Icicle but results unchanged on the customer's hardware.

Here's the question.   customer surrenders the UART in HSS but still uses the CLI to boot UBoot/Linux, customer has no SD card or eMMC so they boot via QSPI NAND flash, need CLI from HSS to do so..   To me if the HSS truly surrenders the UART then use of the CLI is not available, certainly on the Icicle kit this is true.      I ask this since is it possible that with the HSS still having access to the UART that the Console I/O attributed to Linux booting is not available??    I know it's a stretch but wanted to raise the questions.   
thx!

**Comment by griffini** [comment]
Once the HSS surrenders the UART, it will not use it again.  Without UART, then the CLI is not available.  Could you use a GPIO to detect a button press, for instance, as a trigger instead of CLI?

Even if it was using the same UART as Linux, you're likely to see garbled messages rather than Linux's output disappear completely.
